firecracker-control/vsock.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package service
+
+import (
+	"context"
+	"math"
+	"os"
+	"syscall"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	sysCall = syscall.Syscall
+)
+
+// findNextAvailableVsockCID finds first available vsock context ID.
+// It uses VHOST_VSOCK_SET_GUEST_CID ioctl which allows some CID ranges to be statically reserved in advance.
+// The ioctl fails with EADDRINUSE if cid is already taken and with EINVAL if the CID is invalid.
+// Taken from https://bugzilla.redhat.com/show_bug.cgi?id=1291851
+func findNextAvailableVsockCID(ctx context.Context) (uint32, error) {
+	const (
+		// Corresponds to VHOST_VSOCK_SET_GUEST_CID in vhost.h
+		ioctlVsockSetGuestCID = uintptr(0x4008AF60)
+		// 0, 1 and 2 are reserved CIDs, see http://man7.org/linux/man-pages/man7/vsock.7.html
+		startCID        = 3
+		maxCID          = math.MaxUint32
+		vsockDevicePath = "/dev/vhost-vsock"
+	)
+
+	file, err := os.OpenFile(vsockDevicePath, syscall.O_RDWR, 0600)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to open vsock device")
+	}
+
+	defer file.Close()
```

Comment:
(Getting IRL conversation on the record here) When we close this file we lose the reservation we just made, so this function can easily return the same CID if multiple VMs are being spun up in parallel.

Unfortunately, if we just *don't* close the file then when the Firecracker VM tries to use the CID, it will get `EADDRINUSE`. Because there's no way to pass the file descriptor to Firecracker (only the CID number), there's no way to fully eliminate race conditions here.

I think the best we can do right now is
1. Keep the file open right until we start the VM
2. Maintain a map of CIDs the plugin is currently using and don't try to re-use those

That way, the only race condition left is if some other program on the host is also allocating CIDs and it happens to do so with a conflicting CID in the short time frame between when the StartVM API call is made and when Firecracker actually opens the vsock again. I think that should be enough until the Firecracker team implements their new vsock proposal.